### PR TITLE
[IMP] web, base, *: remove cids from URL

### PR DESCRIPTION
**Pull request by jpp-odoo**
*: [hr_holidays, mail, test_mail, test_access_rights, test_main_flows,
exceptions, http, models]

The aim of this commit is to remove the companies IDs from the URL, and
use only the cookies to store the selected companies.
This commit will also improve the handling of access errors. If a user
encounters an access error, if one of its allowed companies has access
to the record, a button to switch to the correct company is shown.

The purpose of this is double: to have more human-readable URLs (part of
the task id : 3557575), and to avoid modifying the current selected
companies when clicking on a shared URL.

Note that this commit also modify the way the companies IDs are encoded
on the cookies. Before this commit, the IDs were separated by a comma
(",") and they were encoded as \054 by the server side but not by the
client side. Now, we use a dash ("-") to have the same encoding in both,
server side and client side.

task-id 3792021

Co-authored-by: Damien Bouvy <dbo@odoo.com>

### CODE REVIEW

File: addons/web/static/tests/legacy/webclient/company_service_tests.js [hunk 1]
```diff
@@ -67,31 +64,3 @@ QUnit.test(
         assert.verifySteps([]);
     }
 );
-
-QUnit.test("extract allowed company ids from url search", async (assert) => {
```
Comment (aab-odoo) [inline addons/web/static/tests/legacy/webclient/company_service_tests.js:71]: cookie

**Review by aab-odoo — APPROVED** [review]
robodoo r+

**Comment by bouvyd** [comment]
@robodoo r-

some stuff to improve before removing this as it will lead to many access errors (notably from email links which currently inject the cid when redirecting the user from the record, which we did not know about)

**Comment by tde-banana-odoo** [comment]
@robodoo r-

Please no.

**Comment by tde-banana-odoo** [comment]
Don't do that: we explicitly give cids in mail to be sure user in MC environment can click on links. We don't want "clean access errors", we want people to be automatically on the right record. That was the purpose of this addition in mail done with @tivisse as we encountered a lot of issues. Don't remove features just to have "nice URLs" please.

**Comment by jpp-odoo** [comment]
> Don't do that: we explicitly give cids in mail to be sure user in MC environment can click on links. We don't want "clean access errors", we want people to be automatically on the right record. That was the purpose of this addition in mail done with @tivisse as we encountered a lot of issues. Don't remove features just to have "nice URLs" please.

This functionality still works, the `_redirect_to_record` function still add the corrects company IDs in the cookies (not in the url). So the user will still be automatically on the right record.

**Comment by tde-banana-odoo** [comment]
Indeed, somehow it seems I landed on original version / WIP ? Maybe a specific temporary commit.

File: addons/test_mail/tests/test_mail_multicompany.py [hunk 2]
```diff
@@ -383,36 +381,34 @@ def test_redirect_to_records_nothread(self):
 
         # when being logged, cids should be based on current user's company unless
         # there is an access issue (not tested here, see 'test_redirect_to_records')
-        self.authenticate(self.user_admin.login, self.user_admin.login)
         for test_record in nothreads:
             for user_company in self.company_admin, self.company_2:
                 with self.subTest(record_name=test_record.name, user_company=user_company):
+                    self.authenticate(self.user_admin.login, self.user_admin.login)
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_multicompany.py:387]: Why this change ? People won't log out log in each time they click on a link anyway ? Tests should still work as before, shouldn't they ?

File: odoo/addons/test_main_flows/tests/test_flow.py [hunk 3]
```diff
@@ -109,9 +109,8 @@ def test_company_switch_access_error(self):
             "action": f"ir.actions.act_window,{act_window.id}",
         })
 
-        current_companies = "%s-%s" % (company1.id, company2.id)
         with mute_logger("odoo.http"):
-            self.start_tour(f"/web#action={act_window.id}&cids={current_companies}", "test_company_switch_access_error", login="admin")
+            self.start_tour(f"/web#action={act_window.id}", "test_company_switch_access_error", login="admin", cookies={"cids": f"{company1.id}-{company2.id}"})
```
Comment (tde-banana-odoo) [inline odoo/addons/test_main_flows/tests/test_flow.py:113]: Note that using cookies will probably quickly lead to having all of them in cookies / a cookie mess. At least URLs were clear about a specific link using some CID

File: addons/test_mail/tests/test_mail_multicompany.py (hunk 2 above)
Comment (jpp-odoo) [reply addons/test_mail/tests/test_mail_multicompany.py:387]: I think that, previous to this commit, the test didn't take into account the modification of the cookies when clicking on a link in a mail (done when loading the webclient based on the url cids). So in each of the subTests when reading from the cookies : https://github.com/odoo/odoo/blob/f71072de8414d84317e6eee2bb0ad2cae25b798c/addons/mail/controllers/mail.py#L74
The cookies are always the same, the ones of the authentication.

Now, as the modification of the cookies is done by the python code, in order to have the same results as before, and to avoid that one subTest have a non-desired effect in another subTest, we need to reset the cookies to the initial ones (after the authentication). So, I placed the authenticate process at the beginning of each subTest.

At least, is what I understood of the test.

File: odoo/addons/base/models/ir_rule.py [hunk 4]
```diff
@@ -243,14 +244,26 @@ def get_record_description(rec):
...
+            suggested_companies = records.sudo()._get_redirect_suggested_company()
+            if suggested_companies and len(suggested_companies) > 1:
+                resolution_info += "\n\n" + _('Note: this might be a multi-company issue. Switching company may help - in Odoo, not in real life!')
+            else:
+                suggested_company = suggested_companies if suggested_companies and suggested_companies in self.env.user.company_ids else None
+                if suggested_company:
+                    context = {'suggested_company': {'id': suggested_company.id, 'display_name': suggested_company.display_name}}
+                    resolution_info += "\n\n" + _('This seems to be a multi-company issue, you might be able to access the record by switching to the company: %s.', suggested_company.display_name)
+                else:
+                    resolution_info += "\n\n" + _('This seems to be a multi-company issue, but you do not have access to the proper company to access the record anyhow.')
```
Comment (ryv-odoo) [inline odoo/addons/base/models/ir_rule.py:258]: IMO, you can simplify these if and handle more correctly the case where there is no company_id ('This seems to be a multi-company issue, but you do not have access to the proper company to access the record anyhow.' => 'Note: this might be a multi-company issue. Switching company may help - in Odoo, not in real life!')

```suggestion
            if len(suggested_companies) != 1:
                resolution_info += "\n\n" + _('Note: this might be a multi-company issue. Switching company may help - in Odoo, not in real life!')
            elif suggested_companies in self.env.user.company_ids:
                context = {'suggested_company': {'id': suggested_company.id, 'display_name': suggested_company.display_name}}
                resolution_info += "\n\n" + _('This seems to be a multi-company issue, you might be able to access the record by switching to the company: %s.', suggested_company.display_name)
            else:
                resolution_info += "\n\n" + _('This seems to be a multi-company issue, but you do not have access to the proper company to access the record anyhow.')
```

File: odoo/addons/test_access_rights/tests/test_feedback.py [hunk 5]
```diff
@@ -398,6 +398,40 @@ def test_warn_company_access(self):
...
+            self.env.invalidate_all()
+            _ = records.with_user(self.user).read(["val"])
+        self.maxDiff = None
+        self.assertEqual(
+            ctx.exception.args[0],
+            """Uh-oh! Looks like you have stumbled upon some top-secret records.
+
+Sorry, %s (id=%s) doesn't have 'read' access to:
+- %s, %s (%s: %s, company=%s)
+ - %s, %s (%s: %s, company=%s)
```
Comment (ryv-odoo) [inline odoo/addons/test_access_rights/tests/test_feedback.py:424]: Can you remove this useless extra begin space here by changing `failing_records = ' \n'...` => `failing_records = '\n'...`.

File: odoo/addons/test_access_rights/tests/test_feedback.py [hunk 6]
```diff
@@ -398,6 +398,40 @@ def test_warn_company_access(self):
...
+
+Blame the following rules:
+- rule 0
+
+If you really, really need access, perhaps you can win over your friendly administrator with a batch of freshly baked cookies.
+
+Note: this might be a multi-company issue. Switching company may help - in Odoo, not in real life!"""
+        % (self.user.name, self.user.id, record_1._description, record_1.display_name, record_1._name, record_1.id, record_1.sudo().company_id.display_name, record_2._description, record_2.display_name, record_2._name, record_2.id, record_2.sudo().company_id.display_name))
```
Comment (ryv-odoo) [inline odoo/addons/test_access_rights/tests/test_feedback.py:432]: Suggestion to clean a little bit the test (variable renaming, update comments, remove noise, f-string, ...):
```suggestion
        """ Test that AccessError handle correctly several companies """
        company_1, company_2 = self.env['res.company'].create([
            {'name': 'Brosse Inc.'},
            {'name': 'Brosse Inc. 2'},
        ])
        records = self.env["test_access_right.some_obj"].create([
            {"val": 1, "company_id": company_1.id},
            {"val": 2, "company_id": company_2.id},
        ])
        record_1, record_2 = records
        self.user.sudo().company_ids = [Command.link(company_1.id), Command.link(company_2.id)]
        self._make_rule('rule 0', "[('company_id', '=', False)]", attr='read')
        self.env.invalidate_all()
        with self.debug_mode(), self.assertRaises(AccessError) as ctx:
            _ = records.with_user(self.user).read(["val"])
        self.assertEqual(
            ctx.exception.args[0],
            f"""Uh-oh! Looks like you have stumbled upon some top-secret records.

Sorry, {self.user.name} (id={self.user.id}) doesn't have 'read' access to:
- {record_1._description}, {record_1.display_name} ({record_1._name}: {record_1.id}, company={record_1.company_id.display_name})
- {record_2._description}, {record_2.display_name} ({record_2._name}: {record_2.id}, company={record_2.company_id.display_name})

Blame the following rules:
- rule 0

If you really, really need access, perhaps you can win over your friendly administrator with a batch of freshly baked cookies.

Note: this might be a multi-company issue. Switching company may help - in Odoo, not in real life!""")
```


Also you should move `self.maxDiff = None` in the `setUpClass` (`cls.maxDiff = None`)

**Review by ryv-odoo — COMMENTED** [review]
Much better, thank you @jpp-odoo . 
I left some suggestions, don't hesitate if you have questions.

**Review by ryv-odoo — APPROVED** [review]
For the base/orm part.
Thanks @jpp-odoo

File: addons/test_mail/tests/test_mail_multicompany.py (hunk 2 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_multicompany.py:387]: Then what happens if you click on multiple links in a given session ? That should probably be a new test that demonstrates the change linked to the new implementation ...

File: addons/web/static/src/views/form/form_error_dialog/redirect_company_dialog.xml [hunk 7]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+  <t t-name="web.RedirectCompanyDialog">
```
Comment (tde-banana-odoo) [inline addons/web/static/src/views/form/form_error_dialog/redirect_company_dialog.xml:3]: :socks:  Strange indent in this file :thinking:

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, few comments from phishing part :) Cheers !

**Comment by ryv-odoo** [comment]
Hello @odoo/rd-security, the rule `PY019` (https://github.com/odoo/enterprise/wiki/security#PY019) isn't correct since https://github.com/odoo/enterprise/pull/55523, right ?

File: addons/web/static/src/views/form/form_controller.js [hunk 8]
```diff
@@ -194,6 +204,18 @@ export class FormController extends Component {
             );
         });
 
+        onError((error) => {
+            const suggestedCompany = error.cause?.data?.context?.suggested_company;
```
Comment (aab-odoo) [inline addons/web/static/src/views/form/form_controller.js:208]: not sure about putting the information in a `context` key. Does it make sense?

File: addons/web/static/src/views/form/form_controller.js [hunk 9]
```diff
@@ -194,6 +204,18 @@ export class FormController extends Component {
...
+            if (error.cause?.data?.name === "odoo.exceptions.AccessError" && suggestedCompany) {
+                const activeCompanyIds = this.companyService.activeCompanyIds;
+                activeCompanyIds.push(suggestedCompany.id);
+                this.companyService.setCompanies(activeCompanyIds, true);
+                return true;
```
Comment (aab-odoo) [inline addons/web/static/src/views/form/form_controller.js:213]: why returning true?

**Review by aab-odoo — APPROVED** [review]

**Review by aab-odoo — APPROVED** [review]
robodoo r+
